### PR TITLE
Git magic dec 13 2021

### PR DESCRIPTION
**Pull request by aeunms**
Culmination of git magic from Nov 25, 2021 - December 8th, 2021

### CODE REVIEW

**Comment by erinha** [comment]
<p>Analyzed commit id : 40faa4046527b26f71574a6ba3aabc9353410298</p><hr><details><summary><b><span style="color: coral">⯅</span> @microsoft/bundle-analysis-app:</b> +1.2 KB</summary><table><tr><th>Metric Name</th><th>Baseline Size</th><th>Compare Size</th><th>Size Diff</th></tr><tr><td>main.js</td><td>59.43 KB</td><td>60.63 KB</td><td><span style="color: coral">⯅</span> +1.2 KB</td></tr><tr><td>Total Size</td><td>59.43 KB</td><td>60.63 KB</td><td><span style="color: coral">⯅</span> +1.2 KB</td></tr></table></details><hr><p>Baseline commit: 1ae1c726af07ed6f4c9f41264f7c48fe81a558d6</p>

File: packages/teams-js/src/internal/internalAPIs.ts
Comment (ndangudubiyyam) [inline packages/teams-js/src/internal/internalAPIs.ts:66]: Why are we returning an error? Isn't it standard practice to throw an Error when it is not supported?

File: packages/teams-js/src/internal/mediaUtil.ts
Comment (ndangudubiyyam) [inline packages/teams-js/src/internal/mediaUtil.ts:64]: Why are we returning an error? Isn't it standard practice to throw an Error when it is not supported?

File: packages/teams-js/src/public/media.ts
Comment (ndangudubiyyam) [inline packages/teams-js/src/public/media.ts:333]: Why are we exporting it even though it is tagged `@internal`?

File: packages/teams-js/src/internal/internalAPIs.ts
Comment (aeunms) [reply packages/teams-js/src/internal/internalAPIs.ts:66]: I'm not entirely sure, but I think they're currently setting up for more changes in this area so I'm thinking it might be best to leave it be since they might have reasons for building their helper in this specific way and it's also internal so we're not expecting anyone else to use it.

Comment (charpercyr) [reply packages/teams-js/src/internal/internalAPIs.ts:66]: I agree with Nikahira, either throwing an error or returning a boolean would make more sense. I am guessing the PR on the main branch contains this change? I am wondering if a PR on the main branch would be appropriate?

Comment (charpercyr) [reply packages/teams-js/src/internal/internalAPIs.ts:66]: I also agree with Nikahira. It should probably either throw an error or return a boolean. I am wondering if a PR on the main branch would be appropriate to fix this?

Comment (ndangudubiyyam) [reply packages/teams-js/src/internal/internalAPIs.ts:66]: I would change it to throw Error. (Do it right the first time). We can always change future releases (dependent functions) to have it called the right way.

File: packages/teams-js/src/public/media.ts
Comment (ndangudubiyyam) [inline packages/teams-js/src/public/media.ts:408]: To have a backward compatibility make it an overloaded function. You can see the selectMedia() method for reference.
The implementation would look something like below
```
   protected notifyEventToHost(mediaEvent: MediaControllerEvent): Promise<void>;
    protected notifyEventToHost(mediaEvent: MediaControllerEvent, callback?: (err?: SdkError) => void): void;
    protected notifyEventToHost(mediaEvent: MediaControllerEvent, callback?: (err?: SdkError) => void): Promise<void> {
      ensureInitialized(FrameContexts.content, FrameContexts.task);
      try {
        isApiSupportedOnMobile(nonFullScreenVideoModeAPISupportVersion);
      } catch (err) {
        const wrappedFunction: InputFunction<void> = () => Promise.reject(err);
        return callCallbackWithSdkErrorFromPromiseAndReturnPromise(wrappedFunction, callback);
      }
      const mediaControllerParams: MediaControllerParam = {
        mediaType: this.getMediaType(),
        mediaControllerEvent: mediaEvent,
      };
      const wrappedFunction: InputFunction<void> = () =>
        Promise.resolve(sendAndHandleSdkError('media.controller', mediaControllerParams));
      return callCallbackWithSdkErrorFromPromiseAndReturnPromise(wrappedFunction, callback);
    }
```

File: packages/teams-js/src/public/media.ts
Comment (ndangudubiyyam) [inline packages/teams-js/src/public/media.ts:424]: Please make an overloaded function like 'notifyEventToHost'

Comment (ndangudubiyyam) [reply packages/teams-js/src/public/media.ts:424]: Also make it protected as the previous methods are protected, unless only stop method is to be accessed by outside classes. I would ask the implementer on what the use case for this.

Comment (aeunms) [reply packages/teams-js/src/public/media.ts:424]: To me it makes sense that the stop() method is public since they're probably planning on adding more public functions like stop() that the client can call which call on the protected methods with specific arguments. I've found a link to the thread on this and messaged it to you, I think it's pretty helpful in understanding the direction of the added media functionalities. If it would still be recommended to use protected, I would love to pick your brain and find out more!

File: packages/teams-js/src/internal/internalAPIs.ts
Comment (aeunms) [reply packages/teams-js/src/internal/internalAPIs.ts:66]: In that case, I think I should reach out to the v1 folks before this is merged so it's consistent across the versions.

File: packages/teams-js/src/public/media.ts
Comment (ndangudubiyyam) [inline packages/teams-js/src/public/media.ts:445]: Won't it be nice to have an overloaded/only constructor to assign the controllerCallback? (Follow up on this issue with v1 teams)

File: packages/teams-js/src/public/meeting.ts
Comment (ndangudubiyyam) [inline packages/teams-js/src/public/meeting.ts:574]: Why are they tagged @hidden ?
It will be helpful for the SDK developers to know what the helper function input/output.

File: packages/teams-js/test/public/media.spec.ts
Comment (ndangudubiyyam) [inline packages/teams-js/test/public/media.spec.ts:444]: Please write a test to have the callback tested too.

File: packages/teams-js/test/public/media.spec.ts
Comment (ndangudubiyyam) [inline packages/teams-js/test/public/media.spec.ts:455]: Junit doesn't test anything. plese use await with async before expect so the test would wait for the promise to be resolved.

File: packages/teams-js/test/public/media.spec.ts
Comment (ndangudubiyyam) [inline packages/teams-js/test/public/media.spec.ts:479]: Junit doesn't test anything. plese use await with async before expect() statement so the test would wait for the promise to be resolved/rejected. Please add an additional callback testcase for this promise.
Please refer sharing.spec.ts for reference.

**Comment by erinha** [comment]
<p>Analyzed commit id : 290c6f392a12d65b1ea56bc6a4370df17a20a4bd</p><hr><details><summary><b><span style="color: coral">⯅</span> @microsoft/bundle-analysis-app:</b> +1.23 KB</summary><table><tr><th>Metric Name</th><th>Baseline Size</th><th>Compare Size</th><th>Size Diff</th></tr><tr><td>main.js</td><td>59.43 KB</td><td>60.66 KB</td><td><span style="color: coral">⯅</span> +1.23 KB</td></tr><tr><td>Total Size</td><td>59.43 KB</td><td>60.66 KB</td><td><span style="color: coral">⯅</span> +1.23 KB</td></tr></table></details><hr><p>Baseline commit: 1ae1c726af07ed6f4c9f41264f7c48fe81a558d6</p>

File: packages/teams-js/src/public/media.ts
Comment (ndangudubiyyam) [inline packages/teams-js/src/public/media.ts:435]: Does this method definition compile??
The moment you pass the callback to the function notifyEventToHost, it will resolve to the return type void instead of Promise<void>
```
public stop(): Promise<void>;
    public stop(callback: (err?: SdkError) => void): void;
    public stop(callback?: (err?: SdkError) => void): Promise<void> {
      return Promise.resolve(this.notifyEventToHost(MediaControllerEvent.StopRecording, callback));
    }
```
  }

File: packages/teams-js/test/public/media.spec.ts
Comment (ndangudubiyyam) [inline packages/teams-js/test/public/media.spec.ts:502]: same as above

File: packages/teams-js/test/public/media.spec.ts
Comment (ndangudubiyyam) [inline packages/teams-js/test/public/media.spec.ts:539]: same as above

File: packages/teams-js/test/public/media.spec.ts
Comment (ndangudubiyyam) [inline packages/teams-js/test/public/media.spec.ts:833]: same comment as above

File: packages/teams-js/src/public/meeting.ts
Comment (aeunms) [reply packages/teams-js/src/public/meeting.ts:574]: The @hidden tag is so the docs don't show up in the public documentation for the clients rather than for the SDK devs.

File: packages/teams-js/test/public/media.spec.ts
Comment (aeunms) [reply packages/teams-js/test/public/media.spec.ts:444]: The example is super helpful, thank you!

File: packages/teams-js/src/public/media.ts
Comment (aeunms) [reply packages/teams-js/src/public/media.ts:435]: I'm still not sure what you mean by this. I thought regardless of the callback, it should call the notifyEventToHost() which returns a promise.

File: packages/teams-js/src/internal/internalAPIs.ts
Comment (aeunms) [reply packages/teams-js/src/internal/internalAPIs.ts:66]: To get unblocked, tabling this issue. Will create a work item.

File: packages/teams-js/src/internal/mediaUtil.ts
Comment (aeunms) [reply packages/teams-js/src/internal/mediaUtil.ts:64]: To get unblocked, tabling this issue. Will create a work item.

File: packages/teams-js/src/public/media.ts
Comment (aeunms) [reply packages/teams-js/src/public/media.ts:457]: To get unblocked, tabling this issue. Will create a work item.

File: packages/teams-js/test/public/media.spec.ts
Comment (aeunms) [reply packages/teams-js/test/public/media.spec.ts:830]: Agree this should be fixed. To get unblocked, tabling this issue. Will create a work item.

File: packages/teams-js/src/internal/internalAPIs.ts
Comment (TrevorJoelHarris) [reply packages/teams-js/src/internal/internalAPIs.ts:66]: @aeunms can you please reply to this and link the bug or work item you open for it?

In order to unblock the merge, we're going to have whoever the current git magician is focus on opening bugs for addressing any complicated changes that come up during merge PRs. We'll let whoever picks those bugs up figure out whether or not the change makes sense in v1 or v2 (v1 seems sensible here)

File: packages/teams-js/src/internal/mediaUtil.ts
Comment (TrevorJoelHarris) [reply packages/teams-js/src/internal/mediaUtil.ts:64]: Same thing: please link whatever you open to track this! Might be worth combining with the similar issue from internalAPI.ts?

File: packages/teams-js/src/public/media.ts
Comment (TrevorJoelHarris) [reply packages/teams-js/src/public/media.ts]: @aeunms How much work is it to add this overload? If it's easy and simple we can do it as part of the merge, if not we can open a bug and assign it out.

File: packages/teams-js/src/public/media.ts
Comment (TrevorJoelHarris) [reply packages/teams-js/src/public/media.ts:435]: @ndangudubiyyam If you haven't already, can you please follow up with @aeunms as soon as possible so we can figure out if we need to open a bug and/or fix this before merging.

File: packages/teams-js/src/public/media.ts
Comment (TrevorJoelHarris) [reply packages/teams-js/src/public/media.ts]: This seems like a good candidate for a bug to follow up with the implementer.

File: packages/teams-js/src/public/media.ts
Comment (aeunms) [reply packages/teams-js/src/public/media.ts]: I missed resolving this, it has already been overloaded.

Comment (aeunms) [reply packages/teams-js/src/public/media.ts]: doing a double take- it looks like actually, I just missed overloading this. Will look at this with fresher eyes.

File: packages/teams-js/test/public/media.spec.ts
Comment (TrevorJoelHarris) [reply packages/teams-js/test/public/media.spec.ts]: If this is easy/straightforward we can do it as part of the merge, but otherwise can open a bug to have someone handle it asynchronously (ha!) from merging.

**Review by TrevorJoelHarris — CHANGES_REQUESTED** [review]
:clock1:

File: packages/teams-js/src/internal/internalAPIs.ts
Comment (aeunms) [reply packages/teams-js/src/internal/internalAPIs.ts:66]: As discussed offline, I am creating a bug item and linking it in our Teams channel.

File: packages/teams-js/src/internal/mediaUtil.ts
Comment (aeunms) [reply packages/teams-js/src/internal/mediaUtil.ts:64]: As discussed offline, I am creating a bug item and linking it in our Teams channel.

File: packages/teams-js/src/public/media.ts
Comment (aeunms) [reply packages/teams-js/src/public/media.ts]: As discussed offline, the promisifying (and subsequently, the overloading) will be done by Niharika here. I have created a bug item and will link to it in our Teams git magic thread.

File: packages/teams-js/test/public/media.spec.ts
Comment (aeunms) [reply packages/teams-js/test/public/media.spec.ts]: Opened a bug item, will link to it in our teams git magic channel

File: packages/teams-js/test/public/media.spec.ts
Comment (aeunms) [reply packages/teams-js/test/public/media.spec.ts]: Opening a bug item and will link to it in our teams git magic channel, thanks!

File: packages/teams-js/test/public/media.spec.ts
Comment (aeunms) [reply packages/teams-js/test/public/media.spec.ts]: Opening a bug item and will link to it in our teams git magic channel, thanks!

File: packages/teams-js/test/public/media.spec.ts
Comment (aeunms) [reply packages/teams-js/test/public/media.spec.ts:500]: Included in a bug item, will link to it in teams git magic channel

File: packages/teams-js/test/public/media.spec.ts
Comment (aeunms) [reply packages/teams-js/test/public/media.spec.ts]: Included in a bug item, will link to it in teams git magic channel

**Review by TrevorJoelHarris — APPROVED** [review]

File: packages/teams-js/src/public/media.ts
Comment (ndangudubiyyam) [reply packages/teams-js/src/public/media.ts]: I thought we are reverting the promise code to v1 version

**Review by ndangudubiyyam — APPROVED** [review]
Merge it

**Pull request merged**
